tests/Plugin/AddPathPluginTest.php

```diff
@@ -0,0 +1,136 @@
+use Psr\Http\Message\RequestInterface;
+
+class AddPathPluginTest extends TestCase
+{
+    /**
+     * @var Plugin
+     */
+    private $plugin;
+
+    /**
+     * @var callable empty
+     */
+    private $callable;
+
+    protected function setUp()
+    {
+        $this->callable = function () {};
+        $this->plugin = new AddPathPlugin(new Uri('/api'));
+    }
+
+    public function testRewriteSameUrl()
+    {
+        $verify = function (RequestInterface $request) {
+            $this->assertEquals('https://example.com/api/foo', $request->getUri()->__toString());
+        };
+
+        $request = new Request('GET', 'https://example.com/foo', ['Content-Type'=>'text/html']);
+        $this->plugin->handleRequest($request, $verify, $this->callable);
+
+        // Make a second call with the same $request object
+        $this->plugin->handleRequest($request, $verify, $this->callable);
+
+        // Make a new call with a new object but same URL
+        $request = new Request('GET', 'https://example.com/foo', ['Content-Type'=>'text/plain']);
+        $this->plugin->handleRequest($request, $verify, $this->callable);
+    }
+
+    public function testRewriteCallingThePluginTwice()
+    {
+        $request = new Request('GET', 'https://example.com/foo');
+        $this->plugin->handleRequest($request, function (RequestInterface $request) {
+            $this->assertEquals('https://example.com/api/foo', $request->getUri()->__toString());
+
+            // Run the plugin again with the modified request
+            $this->plugin->handleRequest($request, function (RequestInterface $request) {
+                $this->assertEquals('https://example.com/api/foo', $request->getUri()->__toString());
+            }, $this->callable);
+        }, $this->callable);
+    }
+
+    public function testRewriteWithDifferentUrl()
+    {
+        $request = new Request('GET', 'https://example.com/foo');
+        $this->plugin->handleRequest($request, function (RequestInterface $request) {
+            $this->assertEquals('https://example.com/api/foo', $request->getUri()->__toString());
+        }, $this->callable);
+
+        $request = new Request('GET', 'https://example.com/bar');
+        $this->plugin->handleRequest($request, function (RequestInterface $request) {
+            $this->assertEquals('https://example.com/api/bar', $request->getUri()->__toString());
+        }, $this->callable);
+    }
+
+    public function testRewriteWithDifferentUrlWhenSecondUrlIncludesAddedPath()
+    {
+        $verify = function (RequestInterface $request) {
+            $this->assertEquals('https://example.com/api/foo', $request->getUri()->__toString());
+        };
+
+        $request = new Request('GET', 'https://example.com/foo');
+        $this->plugin->handleRequest($request, $verify, $this->callable);
+
+        $request = new Request('GET', 'https://example.com/api/foo');
+        $this->plugin->handleRequest($request, $verify, $this->callable);
```

Comment:
so in this case we do not add the prefix, even though we logically should. it seems wrong to me that the behaviour depends on the order of requests. if you reorder the 2, the test will fail because this thing gets rewritten to https://example.com/api/api/foo.